vignettes/make_vocab_and_lookup.R

```diff
@@ -0,0 +1,201 @@
+                                  word_freq_tab = NULL) {
+  unnested_processed_wiktionary <- unnest_pw(processed_wiktionary, clean = TRUE)
+  token_counts <- get_token_counts(unnested_processed_wiktionary, word_freq_tab)
+  
+  # punctuation, numbers, [CLS], etc. from wordpiece vocab
+  fancy_wp_tokens <- get_fancy_wp_tokens(wp_vocab)
+
+  actual_wp_words <- get_actual_wp_words(processed_wiktionary, wp_vocab)
+  # short actual words from wp will get a VIP pass directly to vocab.
+  # Though most of them wouldn't be considered words by most people. :)
+  short_words <- actual_wp_words[nchar(actual_wp_words) <= 3]
+  rest_of_words <- actual_wp_words[nchar(actual_wp_words) > 3]
+  # make sure we keep enough tokens to cover the rest of the wp words
+  processed_tokens <- unnested_processed_wiktionary %>% 
+    dplyr::filter(word %in% rest_of_words) %>% 
+    dplyr::distinct(token) %>% 
+    dplyr::pull(token)
+  
+  # we *didn't* clean the processed tokens before this, so clean them now.
+  # This mainly filters out a few capitalized proper nouns that sneaked back in,
+  # but we'll recover the lowercase version of those later.
+  processed_tokens <- processed_tokens[stringr::str_detect(processed_tokens,
+                                                           pattern = "[^a-z#]", 
+                                                           negate = TRUE)]
+  
+  wp_proper_nouns <- get_wp_proper_nouns(processed_wiktionary, wp_vocab)
+  # There's a lot of overlap with "actual" words (e.g. apple). This 
+  # *shouldn't* make a difference, but just to be sure, filter those out:
+  wp_proper_nouns <- setdiff(wp_proper_nouns, actual_wp_words)
+
+  # still need to add in letters, letter pairs for completeness.
+  letter_pairs <- paste0("##", as.vector(outer(letters, letters, FUN = paste0)))
+  # remove most pairs that end in "s"
+  letter_pairs <- letter_pairs[!grepl("##[^sc]s", letter_pairs)]
+  
+  supplemental_tokens <- c(
+    paste0("##", letters),
+    letters,
+    letter_pairs
+  )
+  
+  vocab <- c(
+    "[PAD]", # make sure this one is first.
+    fancy_wp_tokens,
+    short_words,
+    processed_tokens,
+    wp_proper_nouns,
+    supplemental_tokens
+  ) %>% 
+    unique()
+  
+  if (length(vocab) < target_vocab_size) {
+    # Fill up the difference with top tokens, as judged by rank in token_counts.
+    # Maybe later be more sophisticated about proper nouns.
+    n_tokens_to_add <- target_vocab_size - length(vocab)
+    additional_tokens <- token_counts %>% 
+      dplyr::filter(! token %in% vocab) %>% 
+      dplyr::arrange(rank) %>% 
+      head(n_tokens_to_add) %>% 
+      pull(token)
+    vocab <- unique(c(vocab, additional_tokens))
+  } 
+  lookup <- make_lookup(unnested_processed_wiktionary, vocab)
+  return(list(vocab = vocab, lookup = lookup))
+}
+
+unnest_pw <- function(processed_wiktionary, clean = TRUE) {
+  if (clean) { # keep only words in all lowercase latin alphabet
```

Comment:
Do we maybe want to remove diacritics from the processed wiktionary and then remove these and also remove duplicates? The "and then also duplicates" part complicates things, but otherwise I worry that we might be throwing away info for things like... well, not "resume" but other technically-accented-but-not-in-typed-English words.

Reply:
(piecemaker has a function to do the diacritic removal, btw)

Reply:
yeah, that's a good idea. Curious if you can find an example of such a word in wiktionary.

Reply:
https://en.wiktionary.org/wiki/Category:English_terms_spelled_with_%C3%89

Except wow that list is bad. And no luck so far finding ones that aren't also in wk as their unaccented form. When I have a processor free I can check the dump to find any (I'm thinking something like 

```
dump %>%
  dplyr::filter(stringr::str_detect(word, "[^a-z]")) %>%
  dplyr::mutate(clean_word = piecemaker::remove_diacritics(word))
  dplyr::filter(!(clean_word %in% dump$word))
```

Reply:
This is definitely something to keep in mind for the next iteration of vocab-making. Will comment in code.